packetbeat/protos/dns/dns.go

```diff
@@ -421,165 +450,207 @@ func addDnsToMapStr(m common.MapStr, dns *mkdns.Msg, authority bool, additional
-func rrToMapStr(records []mkdns.RR) []common.MapStr {
+// rrsToMapStr converts an array of RR's to an array of MapStr's.
+func rrsToMapStrs(records []mkdns.RR) []common.MapStr {
 	mapStrArray := make([]common.MapStr, len(records))
 	for i, rr := range records {
 		rrHeader := rr.Header()
-		rrType := rrHeader.Rrtype
 
-		mapStr := common.MapStr{
-			"name":  rrHeader.Name,
-			"type":  dnsTypeToString(rrType),
-			"class": dnsClassToString(rrHeader.Class),
-			"ttl":   strconv.FormatInt(int64(rrHeader.Ttl), 10),
-		}
+		mapStr := rrToMapStr(rr)
+		mapStr["name"] = rrHeader.Name
+		mapStr["type"] = dnsTypeToString(rrHeader.Rrtype)
+		mapStr["class"] = dnsClassToString(rrHeader.Class)
+		mapStr["ttl"] = strconv.FormatInt(int64(rrHeader.Ttl), 10)
 		mapStrArray[i] = mapStr
-
-		switch x := rr.(type) {
-		default:
-			// We don't have special handling for this type
-			logp.Debug("dns", "No special handling for RR type %s", dnsTypeToString(rrType))
-			unsupportedRR := new(mkdns.RFC3597)
-			err := unsupportedRR.ToRFC3597(x)
-			if err == nil {
-				rData, err := hexStringToString(unsupportedRR.Rdata)
-				mapStr["data"] = rData
-				if err != nil {
-					logp.Debug("dns", "%s", err.Error())
-				}
-			} else {
-				logp.Debug("dns", "Rdata for the unhandled RR type %s could not be fetched", dnsTypeToString(rrType))
-			}
-		case *mkdns.A:
-			mapStr["data"] = x.A.String()
-		case *mkdns.AAAA:
-			mapStr["data"] = x.AAAA.String()
-		case *mkdns.CNAME:
-			mapStr["data"] = x.Target
-		case *mkdns.MX:
-			mapStr["preference"] = x.Preference
-			mapStr["data"] = x.Mx
-		case *mkdns.NS:
-			mapStr["data"] = x.Ns
-		case *mkdns.PTR:
-			mapStr["data"] = x.Ptr
-		case *mkdns.RFC3597:
-			// Miekg/dns lib doesn't handle this type
-			//TODO: write a test for this.
-			logp.Debug("dns", "Unknown RR type %s", dnsTypeToString(rrType))
-			rData, err := hexStringToString(x.Rdata)
-			mapStr["data"] = rData
-			if err != nil {
-				logp.Debug("dns", "%s", err.Error())
-			}
-		case *mkdns.SOA:
-			mapStr["rname"] = x.Mbox
-			mapStr["serial"] = x.Serial
-			mapStr["refresh"] = x.Refresh
-			mapStr["retry"] = x.Retry
-			mapStr["expire"] = x.Expire
-			mapStr["minimum"] = x.Minttl
-			mapStr["data"] = x.Ns
-		case *mkdns.SRV:
-			mapStr["priority"] = x.Priority
-			mapStr["weight"] = x.Weight
-			mapStr["port"] = x.Port
-			mapStr["data"] = x.Target
-		case *mkdns.TXT:
-			mapStr["data"] = strings.Join(x.Txt, " ")
-		}
 	}
-
 	return mapStrArray
 }
 
-// dnsQuestionToString converts a Question to a string.
-func dnsQuestionToString(q mkdns.Question) string {
-	name := q.Name
+func rrToString(rr mkdns.RR) string {
+	var st string
 
-	return fmt.Sprintf("class %s, type %s, %s", dnsClassToString(q.Qclass),
-		dnsTypeToString(q.Qtype), name)
+	mapStr := rrToMapStr(rr)
+	data, ok := mapStr["data"]
+	delete(mapStr, "data")
+
+	// Warn: Fields order is not guaranteed...if we want keys could be sorted though...
```

Comment:
Keys are known beforehand but they are numerous and for code maintenance it would be best to only set the keys in one spot, such as in ```rrToMapStr```.

If we want it to be deterministic and sorted, can we consider using [sort.Strings](https://golang.org/pkg/sort/#Strings)?
Of course that is not ideal because it runs yet another loop on the hashMap elements. But it would get the job done and ease RR types maintenance.

I guess the best solution would be to generate code with Golang from a file containing the RR types. But I have never used Golang code generation so I don't see it done in this PR.